File: asherah-go/session_cache_bound_test.go
package asherah_test

import (
	"fmt"
	"os"
	"strings"
	"testing"

	asherah "github.com/godaddy/asherah-ffi/asherah-go"
)

// Verifies that the module-level session cache respects
// SessionCacheMaxSize and uses LRU eviction. Prior to the LRU fix the
// cache used insertion-order FIFO, which evicts hot entries that were
// re-used after insertion.
func TestSessionCache_RoundTripUnderEvictionChurn(t *testing.T) {
	ensureNativeLibrary(t)
	os.Setenv("STATIC_MASTER_KEY_HEX", strings.Repeat("22", 32))

	maxSize := 4
	cfg := asherah.Config{
		ServiceName:         "svc",
		ProductID:           "prod",
		Metastore:           "memory",
		KMS:                 "static",
		SessionCacheMaxSize: &maxSize,
	}
	if err := asherah.Setup(cfg); err != nil {
		t.Fatalf("Setup failed: %v", err)
	}
	defer asherah.Shutdown()

	for i := 0; i < 64; i++ {
		partition := fmt.Sprintf("churn-%d", i)
		payload := []byte(fmt.Sprintf("payload-%d", i))
		ct, err := asherah.Encrypt(partition, payload)
		if err != nil {
			t.Fatalf("Encrypt(%s) failed: %v", partition, err)
		}
		recovered, err := asherah.Decrypt(partition, ct)
		if err != nil {
			t.Fatalf("Decrypt(%s) failed: %v", partition, err)
		}
		if string(recovered) != string(payload) {
			t.Fatalf("partition %s: got %q want %q", partition, recovered, payload)
		}
	}
}

func TestSessionCache_HotPartitionsRoundTripRepeatedly(t *testing.T) {
	ensureNativeLibrary(t)
	os.Setenv("STATIC_MASTER_KEY_HEX", strings.Repeat("22", 32))

	maxSize := 2
	cfg := asherah.Config{
		ServiceName:         "svc",
		ProductID:           "prod",
		Metastore:           "memory",
		KMS:                 "static",
		SessionCacheMaxSize: &maxSize,
	}
	if err := asherah.Setup(cfg); err != nil {
		t.Fatalf("Setup failed: %v", err)
	}
	defer asherah.Shutdown()

	for i := 0; i < 16; i++ {
		ct, err := asherah.Encrypt("hot-a", []byte("a"))
		if err != nil {
			t.Fatalf("Encrypt hot-a: %v", err)
		}
		if r, err := asherah.Decrypt("hot-a", ct); err != nil || string(r) != "a" {
			t.Fatalf("hot-a roundtrip: got %q err=%v", r, err)
		}
		ct, err = asherah.Encrypt("hot-b", []byte("b"))
		if err != nil {
			t.Fatalf("Encrypt hot-b: %v", err)
		}
		if r, err := asherah.Decrypt("hot-b", ct); err != nil || string(r) != "b" {
			t.Fatalf("hot-b roundtrip: got %q err=%v", r, err)
		}
	}
}

func TestSessionCache_DefaultBoundRoundTripsPastThousand(t *testing.T) {
	ensureNativeLibrary(t)
	os.Setenv("STATIC_MASTER_KEY_HEX", strings.Repeat("22", 32))

	cfg := asherah.Config{
		ServiceName: "svc",
		ProductID:   "prod",
		Metastore:   "memory",
		KMS:         "static",
	}
	if err := asherah.Setup(cfg); err != nil {
		t.Fatalf("Setup failed: %v", err)
	}
	defer asherah.Shutdown()

	for i := 0; i < 1100; i++ {
		partition := fmt.Sprintf("default-%d", i)
		payload := []byte(fmt.Sprintf("p%d", i))
		ct, err := asherah.Encrypt(partition, payload)
		if err != nil {
			t.Fatalf("Encrypt(%s) failed: %v", partition, err)
		}
		recovered, err := asherah.Decrypt(partition, ct)
		if err != nil {
			t.Fatalf("Decrypt(%s) failed: %v", partition, err)
		}
		if string(recovered) != string(payload) {
			t.Fatalf("partition %s: got %q want %q", partition, recovered, payload)
		}
	}
}

File: asherah-ruby/test/session_cache_bound_test.rb
require_relative "test_helper"

# Verifies that the module-level session cache respects
# SessionCacheMaxSize. Prior to the LRU/bound fix the cache was
# unbounded — every distinct partition_id touched lived until shutdown.
class SessionCacheBoundTest < Minitest::Test
  def setup
    ENV["STATIC_MASTER_KEY_HEX"] = "22" * 32
  end

  def teardown
    Asherah.shutdown if Asherah.get_setup_status
  rescue Asherah::Error::NotInitialized
    # already shut down
  end

  def base_config(max_size: nil)
    cfg = {
      "ServiceName" => "svc",
      "ProductID" => "prod",
      "Metastore" => "memory",
      "KMS" => "static",
      "EnableSessionCaching" => true,
      "Verbose" => false
    }
    cfg["SessionCacheMaxSize"] = max_size if max_size
    cfg
  end

  def test_round_trip_under_eviction_churn
    Asherah.setup(base_config(max_size: 4))
    32.times do |i|
      partition = "churn-#{i}"
      payload = "payload-#{i}"
      ct = Asherah.encrypt(partition, payload)
      assert_equal payload, Asherah.decrypt(partition, ct).force_encoding("UTF-8")
    end
  end

  def test_hot_partitions_round_trip_repeatedly
    Asherah.setup(base_config(max_size: 2))
    16.times do
      a_ct = Asherah.encrypt("hot-a", "a")
      assert_equal "a", Asherah.decrypt("hot-a", a_ct).force_encoding("UTF-8")
      b_ct = Asherah.encrypt("hot-b", "b")
      assert_equal "b", Asherah.decrypt("hot-b", b_ct).force_encoding("UTF-8")
    end
  end

  def test_default_bound_round_trips_past_thousand
    Asherah.setup(base_config)
    1100.times do |i|
      partition = "default-#{i}"
      payload = "p#{i}"
      ct = Asherah.encrypt(partition, payload)
      assert_equal payload, Asherah.decrypt(partition, ct).force_encoding("UTF-8")
    end
  end

  def test_session_cache_disabled_round_trips
    Asherah.setup(base_config.merge("EnableSessionCaching" => false))
    8.times do |i|
      ct = Asherah.encrypt("nocache-#{i}", "x")
      assert_equal "x", Asherah.decrypt("nocache-#{i}", ct).force_encoding("UTF-8")
    end
  end
end
